### PR TITLE
New add item methods

### PR DESCRIPTION
**Pull request by caalador**
Added method to add
item before or after another item.

---
This change is [Reviewable](https://reviewable.io/reviews/vaadin/flow/8429)

### CODE REVIEW

**Comment by mshabarov** [comment]
___
*[flow-data/src/main/java/com/vaadin/flow/data/provider/AbstractListDataView.java, line 137 at r1](https://reviewable.io/reviews/vaadin/flow/8429#-M8Ktxhb-6O46yeqChHb:-M8Ktxhb-6O46yeqChHc:b-9edrmi) ([raw file](https://github.com/vaadin/flow/blob/a3ece6c8f55572c259ea8ad7ce6a064e43806330/flow-data/src/main/java/com/vaadin/flow/data/provider/AbstractListDataView.java#L137)):*
> ```Java
>     @Override
>     public AbstractListDataView<T> addItemAfter(T item, T after) {
>         final Collection<T> items = getDataProvider().getItems();
> ```

The full data list is used here ignoring filtering and sorting, that means that 'after' item might not be visible for user (it is filtered out), and where does 'after' come from in real life in that case? User can't select it by clicking on it because it is invisible and lies somewhere in backed list. Not it can be retrieved with getNextItem()

File: flow-data/src/main/java/com/vaadin/flow/data/provider/ListDataView.java [hunk 1]
```diff
@@ -89,13 +87,42 @@
...
+     *         item to add
+     * @param after
+     *         item after which to add the item at
+     * @return this ListDataView instance
+     * @throws UnsupportedOperationException
+     *         if backing collection doesn't support modification
+     * @throws IndexOutOfBoundsException
+     *         if given index is outside of item set
+     */
+    V addItemAfter(T item, T after);
```
Comment (pleku) [inline flow-data/src/main/java/com/vaadin/flow/data/provider/ListDataView.java:103]: I was thinking about these methods (and the refresh methods) yesterday but didn't yet create an issue.

Do we have an example of using these in any of the components showing why it is needed over `add()` and then applying sorting ?

**Comment by mshabarov** [comment]
___
*[flow-data/src/main/java/com/vaadin/flow/data/provider/ListDataView.java, line 101 at r1](https://reviewable.io/reviews/vaadin/flow/8429#-M8Kvrbf1pXfZrr-mod0:-M8Kvrbf1pXfZrr-mod1:bnc0sg1) ([raw file](https://github.com/vaadin/flow/blob/a3ece6c8f55572c259ea8ad7ce6a064e43806330/flow-data/src/main/java/com/vaadin/flow/data/provider/ListDataView.java#L101)):*
> ```Java
>      *         if backing collection doesn't support modification
>      * @throws IndexOutOfBoundsException
>      *         if given index is outside of item set
> ```

IllegalArgumentException should be mentioned here?

**Comment by mshabarov** [comment]
___
*[flow-data/src/main/java/com/vaadin/flow/data/provider/AbstractListDataView.java, line 142 at r1](https://reviewable.io/reviews/vaadin/flow/8429#-M8Kw4vn5KIGKbn74t7b:-M8Kw4vn5KIGKbn74t7c:b-sluo2l) ([raw file](https://github.com/vaadin/flow/blob/a3ece6c8f55572c259ea8ad7ce6a064e43806330/flow-data/src/main/java/com/vaadin/flow/data/provider/AbstractListDataView.java#L142)):*
> ```Java
>         }
>         if(items instanceof List) {
>             ((List<T>)items).add(((List<T>) items).indexOf(after)+1, item);
> ```

extract variable for `items` to have only one cast

File: flow-data/src/main/java/com/vaadin/flow/data/provider/AbstractListDataView.java [hunk 2]
```diff
@@ -130,6 +132,35 @@ public boolean isItemPresent(T item) {
...
+        final Collection<T> items = getDataProvider().getItems();
+        if(!items.contains(after)) {
+            throw new IllegalArgumentException("Item to insert after is not available in the data");
+        }
+        if(items instanceof List) {
+            ((List<T>)items).add(((List<T>) items).indexOf(after)+1, item);
+            getDataProvider().refreshAll();
+            return this;
+        }
+        throw new IllegalArgumentException("DataProvider collection is not a list.");
```
Comment (pleku) [inline flow-data/src/main/java/com/vaadin/flow/data/provider/AbstractListDataView.java:146]: Maybe just give the real type be here in the exception ? And for before method too

File: flow-data/src/main/java/com/vaadin/flow/data/provider/ListDataView.java [hunk 3]
```diff
@@ -89,13 +87,42 @@
...
+     *
+     * @param item
+     *         item to add
+     * @param before
+     *         item before which to add the item at
+     * @return this ListDataView instance
+     * @throws UnsupportedOperationException
+     *         if backing collection doesn't support modification
+     * @throws IndexOutOfBoundsException
+     *         if given index is outside of item set
```
Comment (pleku) [inline flow-data/src/main/java/com/vaadin/flow/data/provider/ListDataView.java:116]: given index ??

File: flow-data/src/main/java/com/vaadin/flow/data/provider/ListDataView.java [hunk 4]
```diff
@@ -89,13 +87,42 @@
...
+     *
+     * @param item
+     *         item to add
+     * @param after
+     *         item after which to add the item at
+     * @return this ListDataView instance
+     * @throws UnsupportedOperationException
+     *         if backing collection doesn't support modification
+     * @throws IndexOutOfBoundsException
+     *         if given index is outside of item set
```
Comment (pleku) [inline flow-data/src/main/java/com/vaadin/flow/data/provider/ListDataView.java:101]: given index ??

File: flow-data/src/main/java/com/vaadin/flow/data/provider/ListDataView.java (hunk 1 above)
Comment (caalador) [reply flow-data/src/main/java/com/vaadin/flow/data/provider/ListDataView.java:103]: From gridDemo:
```

File: flow-data/src/main/java/com/vaadin/flow/data/provider/AbstractListDataView.java (hunk 2 above)
Comment (caalador) [reply flow-data/src/main/java/com/vaadin/flow/data/provider/AbstractListDataView.java:146]: added

File: flow-data/src/main/java/com/vaadin/flow/data/provider/ListDataView.java (hunk 3 above)
Comment (caalador) [reply flow-data/src/main/java/com/vaadin/flow/data/provider/ListDataView.java:116]: fixed.

File: flow-data/src/main/java/com/vaadin/flow/data/provider/ListDataView.java (hunk 4 above)
Comment (caalador) [reply flow-data/src/main/java/com/vaadin/flow/data/provider/ListDataView.java:101]: fixed.

File: flow-data/src/main/java/com/vaadin/flow/data/provider/ListDataView.java (hunk 1 above)
Comment (pleku) [reply flow-data/src/main/java/com/vaadin/flow/data/provider/ListDataView.java:103]: Yes that will be simplified nicely

**Comment by mshabarov** [comment]
___
*[flow-data/src/main/java/com/vaadin/flow/data/provider/AbstractListDataView.java, line 137 at r1](https://reviewable.io/reviews/vaadin/flow/8429#-M8Ktxhb-6O46yeqChHb:-M8KzprqCd7t2Qdg0PVg:b-pfbiqb) ([raw file](https://github.com/vaadin/flow/blob/a3ece6c8f55572c259ea8ad7ce6a064e43806330/flow-data/src/main/java/com/vaadin/flow/data/provider/AbstractListDataView.java#L137)):*
<details><summary><i>Previously, caalador wrote…</i></summary><blockquote>

We can never add to the sorted/filtered list by position.
</blockquote></details>

Yes, you're right

**Comment by mshabarov** [comment]
:lgtm:

**Pull request merged**